packages/sync/src/class-sender.php

```diff
@@ -477,51 +478,55 @@ public function do_sync_for_queue( $queue ) {
 			 * @param int    $queue_size The size of the sync queue at the time of processing.
 			 */
 			Settings::set_is_sending( true );
-			$processed_item_ids = apply_filters( 'jetpack_sync_send_data', $items_to_send, $this->codec->name(), microtime( true ), $queue->id, $checkout_duration, $preprocess_duration, $queue->size() );
+			$processed_item_ids = apply_filters( 'jetpack_sync_send_data', $items_to_send, $this->codec->name(), microtime( true ), $queue->id, $checkout_duration, $preprocess_duration, $queue->size(), $buffer->id );
 			Settings::set_is_sending( false );
 		} else {
 			$processed_item_ids = $skipped_items_ids;
 			$skipped_items_ids  = array();
 		}
 
-		if ( ! $processed_item_ids || is_wp_error( $processed_item_ids ) ) {
-			$checked_in_item_ids = $queue->checkin( $buffer );
-			if ( is_wp_error( $checked_in_item_ids ) ) {
-				// phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_error_log
-				error_log( 'Error checking in buffer: ' . $checked_in_item_ids->get_error_message() );
-				$queue->force_checkin();
-			}
-			if ( is_wp_error( $processed_item_ids ) ) {
-				return new \WP_Error( 'wpcom_error', $processed_item_ids->get_error_code() );
-			}
-			// Returning a wpcom_error is a sign to the caller that we should wait a while before syncing again.
-			return new \WP_Error( 'wpcom_error', 'jetpack_sync_send_data_false' );
-		} else {
-			// Detect if the last item ID was an error.
-			$had_wp_error = is_wp_error( end( $processed_item_ids ) );
-			if ( $had_wp_error ) {
-				$wp_error = array_pop( $processed_item_ids );
-			}
-			// Also checkin any items that were skipped.
-			if ( count( $skipped_items_ids ) > 0 ) {
-				$processed_item_ids = array_merge( $processed_item_ids, $skipped_items_ids );
-			}
-			$processed_items = array_intersect_key( $items, array_flip( $processed_item_ids ) );
-			/**
-			 * Allows us to keep track of all the actions that have been sent.
-			 * Allows us to calculate the progress of specific actions.
-			 *
-			 * @since 4.2.0
-			 *
-			 * @param array $processed_actions The actions that we send successfully.
-			 */
-			do_action( 'jetpack_sync_processed_actions', $processed_items );
-			$queue->close( $buffer, $processed_item_ids );
-			// Returning a WP_Error is a sign to the caller that we should wait a while before syncing again.
-			if ( $had_wp_error ) {
-				return new \WP_Error( 'wpcom_error', $wp_error->get_error_code() );
+		if ( 'non-blocking' !== $processed_item_ids ) {
```

Comment:
Do you think it's time to extract this conditional into a new method(s)?

Reply:
I debated pulling it out but for an "experimental/beta" feature I didn't want to introduce large code changes which could have regression concerns.